### PR TITLE
Allow GDExtensions to add editor plugins

### LINKED ISSUE #640: [GDExtension] EditorPlugin Tool
### Godot version

v4.0.dev.custom_build [ed5267f69]

### System information

Manjaro 21 KDE, 1950x threadripper, 1080 TI, Vulkan, 16GB Memory

### Issue description

My problem isn't really an error more or less a missing feature? I'm unable to find a solution to using tool. I found some posts about using the old GDNative and modules. I cannot find the equivalent in GDExtensions if there is any.

### Module solution
`EditorPlugins::add_by_type<MyPlugin>();`

### GDNative solution
`register_tool_class<MyPlugin>();`

### PR DESCRIPTION
**Pull request by dsnopek**
Fixes https://github.com/godotengine/godot-cpp/issues/640

This is an alternative to PR #65592 (but just the GDExtension part) based on some feedback that PR received from the GDExtension team back in December.

It adds two new functions to the `GDExtensionInterface` struct (`editor_add_plugin` and `editor_remove_plugin`) which both take a `StringName` representing a class registered in `ClassDB` (since all GDExtension classes need to be registered in `ClassDB`).

This is used in the companion godot-cpp PR: https://github.com/godotengine/godot-cpp/pull/1114

It allows you to use `EditorPlugins::add_by_type<MyEditorPlugin>()` in your GDExtension, in much the same way as you would in a module.

If this is called during Godot's initialization process, it's just adding the class name to a list, and the actual editor plugins will be created later in `EditorNode::EditorNode()`, in the same place the built-in editor plugins are created. If it's called after the editor is already initialized, then it'll call straight into `EditorNode::get_singleton()->add_extension_editor_plugin()`.

~~This is currently a draft because:~~

- ~~I still need to do the godot-cpp changes and actually test it :-)~~
- ~~If PR #76406 is merged, then this'll need to get reworked to use the new way of registering GDExtension interface functions. That PR is more important, so I'm fine saving this one until that one is resolved.~~

~~For the moment, I'd mainly like to see what others think of this general approach!~~

UPDATE: This is now ready for review!

### CODE REVIEW

**Comment by dsnopek** [comment]
I've rebased this now that PR #76406 has been merged, made the companion godot-cpp PR (https://github.com/godotengine/godot-cpp/pull/1114) and actually tested that this works. :-)

This is now ready for review!

**Comment by BastiaanOlij** [comment]
Discussed this with @dsnopek after last meeting. This seems like a sound solution within the constraints we currently have.

My only remaining remark is that this is worth a wider discussion around our limited access to singletons. The "proper" way for this is that we do have access to the `EditorNode` singleton during registration so we can register the plugin just like core modules do.
We can't because in the **current** approach we're piggy backing singleton access on singletons being registered with the scripting engine and this doesn't happen until all classes are registered.

Imho we should change this part of `godot-cpp` and make the actual `get_singleton()` functions of these classes callable from `godot-cpp`. That is potentially a breaking change on the API but worth it as now we will simply be able to call the correct functions.

cc @vnen

**Comment by dsnopek** [comment]
Thanks, @BastiaanOlij!

Just quick clarification on this point:

> The "proper" way for this is that we do have access to the EditorNode singleton during registration so we can register the plugin just like core modules do.

Core modules actually don't access the `EditorNode` singleton during registration to add their editor plugins, because even for modules, the `EditorNode` singleton won't be created until later.

Modules register a creation function during registration (via `EditorPlugins::add_by_type<>()`), and then in `EditorNode`'s constructor, it calls all the creation functions. This PR does basically the same thing, but instead of registering a creation function, it's registering the name of the class in ClassDB (and then `EditorNode`'s construct creates them just like for modules).

**Comment by BastiaanOlij** [comment]
Ah gotcha, so really in this case we should be use exposing static functions but thats not really possible in this case because it's using a template.

Sigh, some things just don't survive the bridge :) 

Yeah I'm all ok with this going ahead. I just hope we get around to doing the other thing some day too :)

**Comment by anunknowperson** [comment]
I am waiting for this, because until it isn't merged, it is impossible to create advanced plugins with GDExtension (
For now, I have compiled the engine with this PR by myself.

**Comment by anunknowperson** [comment]
@dsnopek also i can't compile this PR with MSVC, I get this error:
.\core/extension/gdextension_interface_dump.gen.h(1589): fatal error C1091: compiler constraint: string length exceeds 65535 bytes

It compiles fine with MinGW. Should it be like this?

**Comment by anunknowperson** [comment]
Also i get this error when i try to port my plugin

at: _editor_init (modules\gltf\register_types.cpp:71)
ERROR: Method 'EditorInterface.get_base_control' has changed and no compatibility fallback has been provided. Please open an issue.
   at: gdextension_classdb_get_method_bind (core\extension\gdextension_interface.cpp:1043)

**Comment by anunknowperson** [comment]
I tried to create simple 3d editor:

test_editor.h

`
#ifndef TEST_EDITOR_PLUGIN_H
#define TEST_EDITOR_PLUGIN_H


#include <godot_cpp/classes/editor_plugin.hpp>
#include <godot_cpp/classes/editor_interface.hpp>
#include <godot_cpp/classes/ref.hpp>
#include <godot_cpp/classes/input_event.hpp>


namespace godot {

class TestEditorPlugin : public EditorPlugin {
	GDCLASSTestEditorPlugin EditorPlugin);


protected:
	void _notification(int p_what);

public:

	static void _bind_methods();

	virtual int _forward_3d_gui_input(Camera3D* p_camera, const Ref<InputEvent>& p_event) override;
	virtual String _get_plugin_name() const override { return "Terrain"; }
	virtual bool _has_main_screen() const override { return false; }
	virtual void _edit(Object* p_object) override;
	virtual bool _handles(Object* p_object) const override;
	virtual void _make_visible(bool p_visible) override;


	TestEditorPlugin ();
	~TestEditorPlugin ();

};


}

#endif 

`

test_editor.cpp contains empty implementations.

When I start a project and create any node, the editor hangs without any errors in the console or anywhere.

**Comment by dsnopek** [comment]
> also i can't compile this PR with MSVC, I get this error:

This was fixed in PR #77137 - I'll rebase this PR to include it in a moment.

> When I start a project and create any node, the editor hangs without any errors in the console or anywhere.

Hm, I did an even simpler test when working on this, and that worked for me. Later, I'll try with your test class and see what happens.

**Comment by anunknowperson** [comment]
@dsnopek Minimal reproduction project:
[src.zip](https://github.com/godotengine/godot/files/11522766/src.zip)

**Comment by anunknowperson** [comment]
Video:

https://github.com/godotengine/godot/assets/54563399/35b24c7d-539d-4519-b1ac-de6d5b4637e0

There are also some errors in the console, but they appear regardless of whether there is any extension or not. (And they don't seem to affect anything, because when no extension is added to the project, nothing is freezes.)

**Comment by dsnopek** [comment]
Thanks! I'm able to reproduce the issue with your MRP.

It seems to be related to the `_handles(Object *)` method: if I remove that method from the class, then the editor doesn't hang.

I'll see if I can debug further...

**Comment by dsnopek** [comment]
@anunknowperson The hanging appears to be a regression caused by this godot-cpp PR: https://github.com/godotengine/godot-cpp/pull/1044 - if I reverse the changes that were committed there, then it doesn't hang, and the editor starts up normally!

For the details: running the MRP in the debugger, it looks like we're hanging on `_instance_binding_mutex.lock()` in `Object::get_instance_binding()`. Stepping up through the stack trace and inspecting all the pointers, in the transition from Godot to godot-cpp we seem to end up in a state where we're treating a random chunk of memory as if it were an `Object`, but it definitely isn't. I think the mutex hangs forever because isn't actually a valid mutex.

A related PR for godot-cpp (https://github.com/godotengine/godot-cpp/pull/1045) caused a similar regression, where it fixed calling normal methods, but broke calling virtual methods. This looks to be the same thing! When calling virtual methods, Godot is passing an `Object **` but godot-cpp (after the afformentioned PR) is treating it as `Object *`.

In any case, that bug is out-of-scope for _this_ PR! As far as I can tell, this PR is correctly registering the editor plugin.

**Comment by anunknowperson** [comment]
@dsnopek thanks!
In other words, to make it work for now, I need to revert #1044? Or will it add some new problems?

**Comment by dsnopek** [comment]
> In other words, to make it work for now, I need to revert https://github.com/godotengine/godot/issues/1044?

Yes.

> Or will it add some new problems?

Well, it did fix a legitimate bug, so you'll be getting that bug back, unfortunately. So, it really depends on if your project runs into that bug too.

The regressions from https://github.com/godotengine/godot/issues/1044 and https://github.com/godotengine/godot/issues/1045 are critical and need to get fixed before Godot 4.1. I'm personally planning to focus my attention on them after the feature freeze next week - until then, we gotta try and finish up any new features (like this editor plugin PR). After the feature freeze, we'll have another 4 weeks to work on bugs. :-)

**Review by BastiaanOlij — APPROVED** [review]
Discussed privately and in the GDExtension meeting. This seems like a clean way to add the ability to register editor plugins properly. Ready to be merged

File: core/extension/gdextension_interface.cpp
Comment (akien-mga) [inline core/extension/gdextension_interface.cpp:46]: `core` should not include `editor`, it should be self-contained. So this needs a different approach.

Comment (dsnopek) [reply core/extension/gdextension_interface.cpp:46]: Ok!

For the case of registering editor plugins during editor startup, this is a really simple change: we just move where the list of class names to register is stored. I only put it on `EditorPlugins` because that's where modules put them. But we can make a special static class just for GDExtensions, and `EditorNode` can reference that.

For the case of registering editor plugins when the editor is already started, it'll be a little messier, because we can't just directly call the method on `EditorNode`. I think I have an idea with passing around function pointers that will work, though.

Update coming soon! :-)

Comment (dsnopek) [reply core/extension/gdextension_interface.cpp]: Fixed in latest push! I added some comments by the function pointers to `EditorNode` so that future contributors know why it's doing that :-)

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**
